lib/PreviewProvider.js

```diff
@@ -0,0 +1,226 @@
+/* globals Task, require*/
+"use strict";
+
+const {Cu} = require("chrome");
+const ss = require("sdk/simple-storage");
+const {setTimeout, clearTimeout} = require("sdk/timers");
+const EMBEDLY_ENDPOINT = require("sdk/simple-prefs").prefs.embedlyEndpoint;
+
+const EMBEDLY_PROXY_MAX_LINKS = 25; // number of links embedly proxy accepts per request
+const CLEAR_CACHE_TIMEOUT = 86400000; // 24 hours to clear cache
+const ALLOWED_QUERY_PARAMS = new Set(["id", "p", "q", "query", "s", "search", "sitesearch"]);
+const ALLOWED_PROTOCOLS = new Set(["http:", "https:"]);
+const DISALLOWED_HOSTS = new Set(["localhost", "127.0.0.1", "0.0.0.0"]);
+const URL_FILTERS = [
+  (item) => !!item.url,
+  (item) => !!(new URL(item.url)),
+  (item) => ALLOWED_PROTOCOLS.has(new URL(item.url).protocol),
+  (item) => !DISALLOWED_HOSTS.has(new URL(item.url).hostname)
+];
+
+Cu.importGlobalProperties(["fetch"]);
+Cu.importGlobalProperties(["URL"]);
+Cu.import("resource://gre/modules/Task.jsm");
+
+function PreviewProvider() {
+  this.init();
+}
+
+PreviewProvider.prototype = {
+  _timeoutID: null,
+
+  /**
+    * Clean up cache on addon bootup
+    */
+  cleanUpCache: Task.async(function*() {
+    let currentTime = Date.now();
+    Object.keys(ss.storage.embedlyData).forEach(item => {
+      if (currentTime - ss.storage.embedlyData[item].accessTime > CLEAR_CACHE_TIMEOUT) {
+        delete ss.storage.embedlyData[item];
+      }
+    });
+    this._cacheTimeout();
+  }),
+
+  /**
+    * Set up the timer for the cache to be cleaned every 24 hours
+    */
+  _cacheTimeout() {
+    this._timeoutID = setTimeout(() => {
+      this.cleanUpCache();
+    }, CLEAR_CACHE_TIMEOUT);
+  },
+
+  /**
+    * Filter out sites that do not have acceptable protocols and hosts
+    */
+  _URLFilter(definition) {
+    return function(item) {
+      let result = true;
+      definition.forEach((test) => {
+        if (!test(item)) {
+          result = false;
+        }
+      });
+      return result;
+    };
+  },
+
+  /**
+    * Santize the URL to remove any unwanted or sensitive information about the link
+    */
+  _sanitizeURL(url) {
+    if (!url) {
+      return "";
+    }
+
+    let queryParams = {};
+    let safeQueryParam = "";
+    let newURL = new URL(url);
+
+    // extract and parse the query parameters, if any
+    if (newURL.search.length !== 0) {
+      let tmp = new Map(newURL.search.slice(1).split("&").map(pair => pair.split("=").map(decodeURIComponent)));
```

Comment:
so this was kinda a tricky and frustrating thing...
queryParams needs to be an actual object, not a map, so that i can filter the allowed query params. so i make a temporary map, but then convert it to an object so i can easily filter across it. 

apparently filtering on a map is not a thing?